optuna/storages/in_memory.py

```diff
@@ -196,174 +229,241 @@ def set_trial_state(self, trial_id, state):
         # type: (int, str, Any) -> None
 
         with self._lock:
-            self.check_trial_is_updatable(trial_id, self.trials[trial_id].state)
+            trial = self._get_trial(trial_id)
+            self.check_trial_is_updatable(trial_id, trial.state)
 
-            trial = copy.copy(self.trials[trial_id])
-            trial.system_attrs = copy.copy(trial.system_attrs)
-            trial.system_attrs[key] = value
-            self.trials[trial_id] = trial
+            with self._lock:
+                self.check_trial_is_updatable(trial_id, trial.state)
+
+                trial = copy.copy(trial)
+                trial.system_attrs = copy.copy(trial.system_attrs)
+                trial.system_attrs[key] = value
+                self._set_trial(trial_id, trial)
 
     def get_trial(self, trial_id):
         # type: (int) -> FrozenTrial
 
         with self._lock:
-            return copy.deepcopy(self.trials[trial_id])
+            return copy.deepcopy(self._get_trial(trial_id))
+
+    def _get_trial(self, trial_id: int) -> FrozenTrial:
+
+        self._check_trial_id(trial_id)
+        study_id, trial_number = self._trial_id_to_study_id_and_number[trial_id]
+        return self._studies[study_id].trials[trial_number]
+
+    def _set_trial(self, trial_id: int, trial: FrozenTrial) -> None:
+        study_id, trial_number = self._trial_id_to_study_id_and_number[trial_id]
+        self._studies[study_id].trials[trial_number] = trial
 
     def get_all_trials(self, study_id, deepcopy=True):
         # type: (int, bool) -> List[FrozenTrial]
 
-        self._check_study_id(study_id)
         with self._lock:
-            if deepcopy:
-                return copy.deepcopy(self.trials)
-            else:
-                return [t for t in self.trials]
+            self._check_study_id(study_id)
+            with self._lock:
+                if deepcopy:
+                    return copy.deepcopy(self._studies[study_id].trials)
+                else:
+                    return self._studies[study_id].trials[:]
 
     def get_n_trials(self, study_id, state=None):
         # type: (int, Optional[TrialState]) -> int
 
-        self._check_study_id(study_id)
-        if state is None:
-            return len(self.trials)
+        with self._lock:
+            self._check_study_id(study_id)
+            if state is None:
+                return len(self._studies[study_id].trials)
 
-        return len([t for t in self.trials if t.state == state])
+            return sum(
+                trial.state == state for trial in self.get_all_trials(study_id, deepcopy=False)
+            )
 
     def _check_study_id(self, study_id):
         # type: (int) -> None
 
-        if study_id != IN_MEMORY_STORAGE_STUDY_ID:
-            raise ValueError(
-                "study_id is supposed to be {} in {}.".format(
-                    IN_MEMORY_STORAGE_STUDY_ID, self.__class__.__name__
-                )
-            )
+        with self._lock:
```

Comment:
Can we skip this lock, assuming that it's acquired by the caller? (Maybe we could do that with `_check_trial_id` as well.)

Reply:
Yes, we should not acquire locks in private methods. I'll fix this.